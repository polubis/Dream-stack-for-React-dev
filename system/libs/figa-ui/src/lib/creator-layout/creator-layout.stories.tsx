import type { Story, Meta } from '@storybook/react';

import { CreatorLayout } from './creator-layout';
import { Box } from '../box';
import { Code } from '../code';
import { Font } from '../font';
import { Button } from '../button';
import { CloseIcon, CodeIcon, PageIcon } from '../icon';

export default {
  component: CreatorLayout,
  title: 'CreatorLayout',
} as Meta;

const Template: Story = () => {
  return (
    <CreatorLayout
      navigation={() => (
        <Box orientation="row" between>
          <Font variant="h5">Article creator</Font>
          <Button shape="rounded" size={1}>
            <CloseIcon />
          </Button>
        </Box>
      )}
      codeToolbox={({ expandPreview, expandBoth }) => (
        <>
          <Button shape="rounded" size={1} onClick={expandPreview}>
            <CodeIcon />
          </Button>
          <Button shape="rounded" size={1} onClick={expandBoth}>
            <PageIcon />
          </Button>
        </>
      )}
      previewToolbox={({ expandCode, expandBoth }) => (
        <>
          <Button shape="rounded" size={1} onClick={expandCode}>
            <CodeIcon />
          </Button>
          <Button shape="rounded" size={1} onClick={expandBoth}>
            <PageIcon />
          </Button>
        </>
      )}
    >
      <Code>
        {`"use strict"
// Abstraction with additional methods.
class List {
  items = [];
 
  constructor(length) {
    this.create(length);
  }
 
  create(length) {
    this.items = Array.from({ length }, (_, i) => i);
  }
   
  remove(idx) {
    this.items = this.items.filter((_, i) => i !== idx);
  }
}
 
// A specific list of users..
class UsersList {
  list = new List(10);
 
  handleRemove(idx) {
    // Removing a list item
    this.list.remove(idx); 
  }
}
 
const usersList = new UsersList();
usersList.handleRemove(9);
// Returns 9 elements.
console.log(usersList.list.items);"use strict"
// Abstraction with additional methods.
class List {
  items = [];
 
  constructor(length) {
    this.create(length);
  }
 
  create(length) {
    this.items = Array.from({ length }, (_, i) => i);
  }
   
  remove(idx) {
    this.items = this.items.filter((_, i) => i !== idx);
  }
}
 
// A specific list of users..
class UsersList {
  list = new List(10);
 
  handleRemove(idx) {
    // Removing a list item
    this.list.remove(idx); 
  }
}
 
const usersList = new UsersList();
usersList.handleRemove(9);
// Returns 9 elements.
console.log(usersList.list.items);"use strict"
// Abstraction with additional methods.
class List {
  items = [];
 
  constructor(length) {
    this.create(length);
  }
 
  create(length) {
    this.items = Array.from({ length }, (_, i) => i);
  }
   
  remove(idx) {
    this.items = this.items.filter((_, i) => i !== idx);
  }
}
 
// A specific list of users..
class UsersList {
  list = new List(10);
 
  handleRemove(idx) {
    // Removing a list item
    this.list.remove(idx); 
  }
}
 
const usersList = new UsersList();
usersList.handleRemove(9);
// Returns 9 elements.
console.log(usersList.list.items);`}
      </Code>
      <div>
        <Font variant="h5">Layout goes here</Font>
      </div>
    </CreatorLayout>
  );
};

export const Default = Template.bind({});
Default.args = {};
