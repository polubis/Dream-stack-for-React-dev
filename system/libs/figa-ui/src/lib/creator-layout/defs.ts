import type { ElementSizeState } from '@system/figa-hooks';
import type { ReactNode, DetailedHTMLProps, HTMLAttributes } from 'react';

type CreatorLayoutHTMLElementProps = Omit<
  DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
  'children'
>;

type CreatorLayoutView =
  | 'undetected'
  | 'both'
  | 'code'
  | 'preview'
  | 'code-full'
  | 'preview-full';

interface CreatorLayoutPayload {
  expandCode: () => void;
  expandPreview: () => void;
  expandBoth: () => void;
  view: CreatorLayoutView;
  size: ElementSizeState;
}

interface CreatorLayoutProps extends CreatorLayoutHTMLElementProps {
  children: [ReactNode, ReactNode];
  codeToolbox: (payload: CreatorLayoutPayload) => ReactNode;
  previewToolbox: (payload: CreatorLayoutPayload) => ReactNode;
  navigation: (payload: CreatorLayoutPayload) => ReactNode;
}

export type {
  CreatorLayoutProps,
  CreatorLayoutHTMLElementProps,
  CreatorLayoutPayload,
  CreatorLayoutView,
};
