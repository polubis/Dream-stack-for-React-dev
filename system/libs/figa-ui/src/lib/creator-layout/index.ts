export * from './creator-layout';
export * from './defs';
