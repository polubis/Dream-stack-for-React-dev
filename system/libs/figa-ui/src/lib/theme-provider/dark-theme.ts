import type { Theme } from './defs';
import { tokens } from './tokens';

const dark: Theme = {
  outline: {
    color: tokens.common.white,
  },
  switch: {
    bg: tokens.gray[50],
    color: tokens.common.white,
    active: {
      bg: tokens.green[200],
    },
  },
  scroll: {
    thumb: tokens.gray[600],
    track: tokens.gray[550],
    thumbHover: tokens.gray[650],
  },
  divider: {
    default: {
      bg: tokens.common.white,
    },
    primary: {
      bg: tokens.primary[100],
    },
  },
  radio: {
    borderColor: tokens.common.white,
    active: {
      bg: tokens.green[200],
      borderColor: tokens.green[200],
    },
  },
  checkbox: {
    borderColor: tokens.common.white,
    checked: {
      bg: tokens.primary[50],
      borderColor: tokens.primary[50],
      color: tokens.common.white,
    },
  },
  tabs: {
    filled: {
      bg: tokens.gray[500],
      active: {
        borderColor: tokens.green[50],
      },
    },
  },
  chips: {
    color: tokens.common.white,
    borderColor: tokens.common.white,
    active: {
      color: tokens.green[50],
      borderColor: tokens.green[50],
    },
  },
  badge: {
    outlined: {
      primary: {
        color: tokens.primary[50],
        borderColor: tokens.primary[50],
      },
      secondary: {
        color: tokens.primary[100],
        borderColor: tokens.primary[100],
      },
      ok: {
        color: tokens.green[100],
        borderColor: tokens.green[100],
      },
      casual: {
        color: tokens.gray[150],
        borderColor: tokens.gray[150],
      },
    },
    filled: {
      primary: {
        color: tokens.common.black,
        bg: tokens.primary[50],
      },
      secondary: {
        color: tokens.common.black,
        bg: tokens.primary[100],
      },
      ok: {
        color: tokens.common.black,
        bg: tokens.green[100],
      },
      casual: {
        color: tokens.common.black,
        bg: tokens.gray[150],
      },
    },
  },
  thumbnail: {
    bg: tokens.backdrop[100],
    color: tokens.common.white,
  },
  filePicker: {
    outline: {
      color: tokens.primary[50],
    },
    invalid: {
      color: tokens.red[150],
      outline: tokens.red[150],
    },
  },
  field: {
    invalid: {
      color: tokens.red[100],
    },
    hint: {
      color: tokens.common.white,
    },
  },
  font: {
    default: {
      color: tokens.common.white,
    },
    primary: {
      color: tokens.primary[100],
    },
  },
  body: {
    bg: tokens.common.black,
  },
  alert: {
    filled: {
      info: {
        bg: tokens.gray[150],
        color: tokens.common.black,
      },
      ok: {
        bg: tokens.green[150],
        color: tokens.common.black,
      },
      error: {
        bg: tokens.red[50],
        color: tokens.common.white,
      },
      warn: {
        bg: tokens.yellow[50],
        color: tokens.common.black,
      },
    },
    outlined: {
      info: {
        color: tokens.common.white,
      },
      ok: {
        color: tokens.green[150],
      },
      error: {
        color: tokens.red[150],
      },
      warn: {
        color: tokens.yellow[50],
      },
    },
  },
  modal: {
    bg: tokens.common.black,
    border: tokens.dark[50],
    backdrop: tokens.backdrop[50],
  },
  button: {
    filled: {
      primary: {
        color: tokens.common.black,
        bg: tokens.primary[50],
        outlineColor: tokens.common.white,
      },
      secondary: {
        color: tokens.common.black,
        bg: tokens.primary[100],
        outlineColor: tokens.common.white,
      },
      tertiary: {
        color: tokens.common.black,
        bg: tokens.common.white,
        outlineColor: tokens.common.white,
      },
    },
    outlined: {
      primary: {
        color: tokens.primary[50],
        bg: 'transparent',
        borderColor: tokens.primary[50],
        outlineColor: tokens.common.white,
      },
      secondary: {
        color: tokens.primary[100],
        bg: 'transparent',
        borderColor: tokens.primary[100],
        outlineColor: tokens.common.white,
      },
      tertiary: {
        color: tokens.common.white,
        bg: 'transparent',
        borderColor: tokens.common.white,
        outlineColor: tokens.common.white,
      },
    },
    ghost: {
      hoverBg: tokens.dark[250],
    },
  },
  select: {
    bg: tokens.dark[50],
    color: tokens.common.white,
    placeholderColor: tokens.gray[50],
    hoverBg: tokens.gray[100],
    optionSeparator: tokens.gray[100],
    optionActiveColor: tokens.secondary[50],
  },
  nav: {
    bg: tokens.dark[100],
    borderColor: tokens.gray[300],
    bgWithOpacity: tokens.dark[200],
  },
  logo: {
    text: {
      first: tokens.purple[100],
      second: tokens.common.white,
      last: tokens.secondary[200],
    },
    bg: {
      first: {
        from: tokens.red[50],
        to: tokens.primary[250],
      },
      second: {
        from: tokens.secondary[250],
        to: tokens.primary[250],
      },
      last: {
        from: tokens.blue[200],
        to: tokens.primary[250],
      },
    },
  },
  progressCircle: {
    bg: tokens.dark[50],
  },
  input: {
    filled: {
      bg: tokens.dark[50],
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    outlined: {
      borderColor: tokens.common.white,
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    empty: {
      color: tokens.common.white,
      placeholderColor: tokens.gray[50],
      fxColor: tokens.common.white,
    },
    invalid: tokens.red[150],
  },
  placeholder: {
    filled: {
      bg: tokens.gray[300],
    },
    outlined: {
      bg: tokens.common.black,
      borderColor: tokens.gray[300],
    },
  },
  box: {
    filled: {
      bg: tokens.gray[300],
    },
    outlined: {
      bg: tokens.common.black,
      borderColor: tokens.gray[300],
    },
  },
  footer: {
    bg: tokens.common.black,
    borderColor: tokens.gray[300],
  },
  code: {
    bg: tokens.dark[150],
    gutters: tokens.dark[150],
    selectionBg: '#4d4d4d',
    foldPlaceholder: '#4d4d4d',
    tags: [
      { tag: 'angleBracket', color: '#da6bad' },
      { tag: 'arithmeticOperator', color: '#fff' },
      { tag: 'bitwiseOperator', color: '#fff' },
      { tag: 'blockComment', color: '#6A9955' },
      { tag: 'bool', color: '#569cd6' },
      { tag: 'brace', color: '#ffd70f' },
      { tag: 'bracket', color: '#da6bad' },
      { tag: 'className', color: '#3abfb0' },
      { tag: 'comment', color: '#6A9955' },
      { tag: 'compareOperator', color: '#fff' },
      { tag: 'controlKeyword', color: '#eba0d9' },
      { tag: 'definition', color: '#4fc1ff' },
      { tag: 'definitionKeyword', color: '#3981c5' },
      { tag: 'definitionOperator', color: '#fff' },
      { tag: 'derefOperator', color: '#fff' },
      { tag: 'function', color: '#dcdc9f' },
      { tag: 'keyword', color: '#3981c5' },
      { tag: 'labelName', color: '#4fc1ff' },
      { tag: 'lineComment', color: '#6A9955' },
      { tag: 'literal', color: '#ef8c59' },
      { tag: 'logicOperator', color: '#fff' },
      { tag: 'moduleKeyword', color: '#eba0d9' },
      { tag: 'name', color: '#4fc1ff' },
      { tag: 'null', color: '#569cd6' },
      { tag: 'number', color: '#accea8' },
      { tag: 'operatorKeyword', color: '#3abfb0' },
      { tag: 'paren', color: '#da6bad' },
      { tag: 'typeName', color: '#3abfb0' },
    ],
  },
  codeBlock: {
    header: {
      bg: tokens.gray[350],
      dots: {
        first: {
          bg: tokens.green[50],
        },
        second: {
          bg: tokens.yellow[50],
        },
        third: {
          bg: tokens.blue[650],
        },
      },
    },
    content: {
      bg: tokens.dark[150],
    },
  },
  list: {
    marker: {
      bg: tokens.primary[50],
    },
  },
  blockquote: {
    bg: tokens.gray[450],
    borderColor: tokens.gray[300],
  },
  loader: {
    primary: tokens.common.white,
    secondary: tokens.primary[50],
  },
  avatars: {
    restCounter: {
      bg: tokens.primary[50],
      color: tokens.common.black,
    },
  },
  creatorLayout: {
    borderColor: tokens.gray[300],
    bg: tokens.common.black,
  },
};

export { dark };
