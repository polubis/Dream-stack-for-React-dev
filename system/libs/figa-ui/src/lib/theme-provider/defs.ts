import type { ReactNode } from 'react';

interface ThemeProviderProps {
  children: ReactNode;
}

interface Theme {
  font: {
    color: string;
  };
  background: {
    body: string;
  };
}

interface Tokens {
  [key: string]: {
    [key: string]: string;
  };
}

type ThemeKey = 'dark' | 'light';
type Themes = Record<ThemeKey, Theme>;
type ThemesList = [ThemeKey, Theme][];

interface ThemeProviderValue {
  key: ThemeKey;
  theme: Theme;
  themes: Themes;
  themesList: ThemesList;
  setTheme: (key: ThemeKey) => void;
}

type UnsafeThemeProviderValue = ThemeProviderValue | null;

type GetDefaultValue = (
  value?: Partial<Pick<ThemeProviderValue, 'key' | 'setTheme'>>
) => ThemeProviderValue;

export type {
  ThemeProviderProps,
  Theme,
  ThemeProviderValue,
  UnsafeThemeProviderValue,
  ThemeKey,
  Tokens,
  Themes,
  ThemesList,
  GetDefaultValue,
};
