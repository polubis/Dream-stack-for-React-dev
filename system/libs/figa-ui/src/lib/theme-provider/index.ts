export * from './theme-provider';
export * from './defs';
