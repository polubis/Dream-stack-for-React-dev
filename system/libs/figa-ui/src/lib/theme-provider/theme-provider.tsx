import { useState, createContext, useMemo, useContext } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import type {
  ThemeKey,
  ThemeProviderProps,
  ThemeProviderValue,
  ThemesList,
  UnsafeThemeProviderValue,
  GetDefaultValue,
} from './defs';
import { themes } from './themes';
import { GlobalStyle } from './global-style';

const Ctx = createContext<UnsafeThemeProviderValue>(null);

const defaultKey = 'dark' as ThemeKey;

const getDefaultValue: GetDefaultValue = ({
  key = defaultKey,
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  setTheme = () => {},
} = {}) => ({
  key,
  setTheme,
  themes,
  theme: themes[key],
  themesList: Object.entries(themes) as ThemesList,
});

const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [key, setKey] = useState(defaultKey);

  const value: ThemeProviderValue = useMemo(
    () =>
      getDefaultValue({
        key,
        setTheme: setKey,
      }),
    [key]
  );

  return (
    <Ctx.Provider value={value}>
      <StyledThemeProvider theme={value.theme}>
        <GlobalStyle />
        {children}
      </StyledThemeProvider>
    </Ctx.Provider>
  );
};

const useThemeProvider = (): ThemeProviderValue => {
  const ctx = useContext(Ctx);

  return ctx ? ctx : getDefaultValue();
};

export { ThemeProvider, useThemeProvider };
