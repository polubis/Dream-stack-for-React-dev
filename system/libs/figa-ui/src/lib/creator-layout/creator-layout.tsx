import styled, { css } from 'styled-components';
import type {
  CreatorLayoutPayload,
  CreatorLayoutProps,
  CreatorLayoutView,
} from './defs';
import { T_DOWN, T_UP, isTDown, isTUp, tokens } from '../theme-provider';
import { slideIn, column, row } from '../shared';
import { useEffect, useState } from 'react';
import c from 'classnames';
import { useElementSize } from '@system/figa-hooks';

const header_height = tokens.spacing[1000];
const toolbox_width = tokens.spacing[700];

const toolbox = (area: string) => css`
  ${column()}
  align-items: center;
  grid-area: ${area};
  padding: ${tokens.spacing[200]} 0;

  & > *:not(:last-child) {
    margin-bottom: ${tokens.spacing[150]};
  }
`;

const content = (area: string) => css`
  grid-area: ${area};
  max-height: calc(100vh - ${header_height});
  overflow-y: auto;
`;

const CodeWrapper = styled.div`
  ${content('code')}
  position: relative;
`;

const PreviewWrapper = styled.div`
  ${content('preview')}
  padding: ${tokens.spacing[250]};
  position: relative;
`;

const CodeToolboxWrapper = styled.div`
  ${toolbox('codeToolbox')}
  border-right: 1px solid ${(props) => props.theme.creatorLayout.borderColor};
`;

const PreviewToolboxWrapper = styled.div`
  ${toolbox('previewToolbox')}
`;

const NavigationWrapper = styled.div`
  ${row()}
  padding: 0 ${tokens.spacing[150]} 0 ${tokens.spacing[200]};
  grid-area: nav;
  background: ${(props) => props.theme.creatorLayout.bg};
  border-bottom: 1px solid ${(props) => props.theme.creatorLayout.borderColor};

  & > * {
    width: 100%;
  }
`;

const Container = styled.div`
  position: relative;
  display: grid;
  grid-template-columns: 1fr ${toolbox_width} 1fr ${toolbox_width};
  grid-template-rows: ${header_height} 1fr;
  grid-template-areas:
    'nav nav nav nav'
    'code codeToolbox preview previewToolbox';
  min-height: 100vh;

  @media ${T_UP} {
    &.code-full {
      grid-template-areas:
        'nav nav nav nav'
        'code code code codeToolbox';
    }

    &.preview-full {
      grid-template-areas:
        'nav nav nav nav'
        'preview preview preview previewToolbox';
    }
  }

  @media ${T_DOWN} {
    grid-template-columns: 1fr ${toolbox_width};
    grid-template-rows: ${header_height} 1fr;
    grid-template-areas:
      'nav nav'
      'content toolbox';

    ${CodeWrapper} {
      ${slideIn('0px', '0px')}
      grid-area: content;
    }

    ${CodeToolboxWrapper} {
      ${slideIn('0px', '0px')}
      grid-area: toolbox;
    }

    ${PreviewWrapper} {
      ${slideIn('0px', '0px')}
      grid-area: content;
    }

    ${PreviewToolboxWrapper} {
      ${slideIn('0px', '0px')}
      grid-area: toolbox;
    }
  }
`;

const CreatorLayout = ({
  className,
  children,
  previewToolbox,
  codeToolbox,
  navigation,
  ...props
}: CreatorLayoutProps) => {
  const [view, setView] = useState<CreatorLayoutView>('undetected');
  const { state: size } = useElementSize({ delay: 20 });

  const [Code, Preview] = children;

  const expandCode = (): void => {
    if (size.status !== 'detected') return;

    const { width } = size;

    if (isTDown(width)) {
      setView('code');
      return;
    }

    setView('code-full');
  };

  const expandPreview = (): void => {
    if (size.status !== 'detected') return;

    const { width } = size;

    if (isTDown(width)) {
      setView('preview');
      return;
    }

    setView('preview-full');
  };

  const expandBoth = (): void => {
    if (size.status !== 'detected') return;

    const { width } = size;

    if (!isTUp(width)) return;

    setView('both');
  };

  useEffect(() => {
    if (size.status !== 'detected') return;

    const { width } = size;

    if (isTDown(width)) {
      setView('code');
      return;
    }

    if (isTUp(width)) {
      setView('both');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [size]);

  const payload: CreatorLayoutPayload = {
    expandCode,
    expandPreview,
    expandBoth,
    view,
    size,
  };

  return (
    <Container
      className={c('creator-layout', view, className)}
      {...(props as Record<string, unknown>)}
    >
      <NavigationWrapper>{navigation(payload)}</NavigationWrapper>

      {view !== 'undetected' && (
        <>
          {(view === 'both' || view === 'code' || view === 'code-full') && (
            <>
              <CodeWrapper>{Code}</CodeWrapper>
              <CodeToolboxWrapper>{codeToolbox(payload)}</CodeToolboxWrapper>
            </>
          )}
          {(view === 'both' ||
            view === 'preview' ||
            view === 'preview-full') && (
            <>
              <PreviewWrapper>{Preview}</PreviewWrapper>
              <PreviewToolboxWrapper>
                {previewToolbox(payload)}
              </PreviewToolboxWrapper>
            </>
          )}
        </>
      )}
    </Container>
  );
};

export { CreatorLayout };
