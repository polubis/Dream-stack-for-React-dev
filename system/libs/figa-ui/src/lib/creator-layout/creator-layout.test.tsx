import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { CreatorLayout } from './creator-layout';
import { ThemeProvider } from '../theme-provider';
import type { CreatorLayoutProps } from './defs';

describe('Creator layout can be used when: ', () => {
  const HEIGHT = 600;
  const WIDTH = 800;

  const creatorLayoutFixture = (props?: Partial<CreatorLayoutProps>) => {
    const result = render(
      <ThemeProvider>
        <CreatorLayout
          navigation={() => <button>Navigation</button>}
          previewToolbox={() => <button>Preview Toolbox Button</button>}
          codeToolbox={() => <button>Code Toolbox Button</button>}
          {...props}
        >
          <div>Code content</div>
          <div>Preview content</div>
        </CreatorLayout>
      </ThemeProvider>
    );

    return result;
  };

  const resizeObserverFixture = (height = HEIGHT, width = WIDTH) => {
    const originalResizeObserver = global.ResizeObserver;
    const disconnectSpy = jest.fn();
    const observeSpy = jest.fn();

    const ResizeObserverMock = class MockedResizeObserver {
      constructor(cb: ResizeObserverCallback) {
        setTimeout(() => {
          cb(
            [
              {
                contentRect: {
                  height,
                  width,
                },
              },
            ] as ResizeObserverEntry[],
            this
          );
        }, 150);
      }

      // eslint-disable-next-line @typescript-eslint/no-empty-function
      observe = observeSpy;

      // eslint-disable-next-line @typescript-eslint/no-empty-function
      unobserve = () => {};

      // eslint-disable-next-line @typescript-eslint/no-empty-function
      disconnect = disconnectSpy;
    };

    return {
      disconnectSpy,
      observeSpy,
      mock: () => {
        global.ResizeObserver = ResizeObserverMock;
      },
      restore: () => {
        global.ResizeObserver = originalResizeObserver;
      },
    };
  };

  it('[FRAGILE] assigns classes', () => {
    const { mock, restore } = resizeObserverFixture();

    mock();

    const { container, asFragment } = creatorLayoutFixture({
      className: 'my-class',
    });

    const component = container.querySelector('.creator-layout');

    expect(component?.className).toContain(
      'creator-layout undetected my-class'
    );
    expect(asFragment()).toMatchSnapshot();

    restore();
  });

  it('skips content rendering if view is undetected', () => {
    const { mock, restore } = resizeObserverFixture();

    mock();

    creatorLayoutFixture();

    screen.getByText(/Navigation/);
    expect(screen.queryByText(/Preview Toolbox Button/)).toBeFalsy();

    restore();
  });

  it('renders preview, code and their toolboxes if there is enough width', async () => {
    const { mock, restore } = resizeObserverFixture(HEIGHT, 1500);

    mock();

    const { asFragment } = creatorLayoutFixture();

    await waitFor(() => {
      screen.getByText(/Preview Toolbox Button/);
      screen.getByText(/Code Toolbox Button/);
      screen.getByText(/Code content/);
      screen.getByText(/Preview content/);
    });

    expect(asFragment()).toMatchSnapshot();

    restore();
  });

  it('renders code and code toolbox by default on smaller devices', async () => {
    const { mock, restore } = resizeObserverFixture(HEIGHT, 500);

    mock();

    const { asFragment } = creatorLayoutFixture();

    await waitFor(() => {
      screen.getByText(/Code Toolbox Button/);
      screen.getByText(/Code content/);
      expect(screen.queryByText(/Preview content/)).toBeFalsy();
      expect(screen.queryByText(/Preview Toolbox Button/)).toBeFalsy();
    });

    expect(asFragment()).toMatchSnapshot();

    restore();
  });

  it('renders different views and allows to change view', async () => {
    const { mock, restore } = resizeObserverFixture(HEIGHT, 1500);

    mock();

    creatorLayoutFixture({
      navigation: ({ expandCode, expandBoth, expandPreview }) => (
        <>
          <button onClick={expandCode}>Expand code</button>
          <button onClick={expandBoth}>Expand both</button>
          <button onClick={expandPreview}>Expand preview</button>
        </>
      ),
    });

    await waitFor(() => {
      screen.getByText(/Expand code/);
      screen.getByText(/Expand both/);
      screen.getByText(/Expand preview/);
      screen.getByText(/Code content/);
      screen.getByText(/Preview content/);
      screen.getByText(/Preview Toolbox Button/);
    });

    fireEvent.click(screen.getByText(/Expand code/));

    await waitFor(() => {
      screen.getByText(/Code content/);
      expect(screen.queryByText(/Preview content/)).toBeFalsy();
      expect(screen.queryByText(/Preview Toolbox Button/)).toBeFalsy();
    });

    fireEvent.click(screen.getByText(/Expand preview/));

    await waitFor(() => {
      screen.getByText(/Preview content/);
      expect(screen.queryByText(/Code content/)).toBeFalsy();
      expect(screen.queryByText(/Code Toolbox Button/)).toBeFalsy();
    });

    fireEvent.click(screen.getByText(/Expand both/));

    await waitFor(() => {
      screen.getByText(/Expand code/);
      screen.getByText(/Expand both/);
      screen.getByText(/Expand preview/);
      screen.getByText(/Code content/);
      screen.getByText(/Preview content/);
      screen.getByText(/Preview Toolbox Button/);
    });

    restore();
  });
});
