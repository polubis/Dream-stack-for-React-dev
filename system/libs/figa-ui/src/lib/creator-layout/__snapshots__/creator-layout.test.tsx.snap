// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creator layout can be used when:  [FRAGILE] assigns classes 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER DFtOs creator-layout code my-class"
  >
    <div
      class="sc-eDvSVe kyLWFC"
    >
      <button>
        Navigation
      </button>
    </div>
    <div
      class="sc-bcXHqe CwXFH"
    >
      <div>
        Code content
      </div>
    </div>
    <div
      class="sc-dkrFOg fPnWkf"
    >
      <button>
        Code Toolbox Button
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Creator layout can be used when:  renders code and code toolbox by default on smaller devices 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER DFtOs creator-layout code"
  >
    <div
      class="sc-eDvSVe kyLWFC"
    >
      <button>
        Navigation
      </button>
    </div>
    <div
      class="sc-bcXHqe CwXFH"
    >
      <div>
        Code content
      </div>
    </div>
    <div
      class="sc-dkrFOg fPnWkf"
    >
      <button>
        Code Toolbox Button
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Creator layout can be used when:  renders preview, code and their toolboxes if there is enough width 1`] = `
<DocumentFragment>
  <div
    class="sc-jSUZER DFtOs creator-layout both"
  >
    <div
      class="sc-eDvSVe kyLWFC"
    >
      <button>
        Navigation
      </button>
    </div>
    <div
      class="sc-bcXHqe CwXFH"
    >
      <div>
        Code content
      </div>
    </div>
    <div
      class="sc-dkrFOg fPnWkf"
    >
      <button>
        Code Toolbox Button
      </button>
    </div>
    <div
      class="sc-gswNZR etFxWQ"
    >
      <div>
        Preview content
      </div>
    </div>
    <div
      class="sc-hLBbgP jPnqra"
    >
      <button>
        Preview Toolbox Button
      </button>
    </div>
  </div>
</DocumentFragment>
`;
